# Sampling Assignment
### Name: Harshita Pandey
### Roll No: 102003235
### Group No: CO10


## Importing the dataset

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv('Creditcard_data.csv')

In [3]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      1  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
df.shape

(772, 31)

In [5]:
freq = df["Class"].value_counts(normalize=True)
print(freq)
print(freq*100)

0    0.988342
1    0.011658
Name: Class, dtype: float64
0    98.834197
1     1.165803
Name: Class, dtype: float64


The given dataset is highly imbalanced. From the above calculations, we find out that the data is 98.83% imabalanced.

Since the dataset is highly imabalanced, we use the SMOTE or the Synthetic Minority Oversampling Technique

In [6]:
pip install imblearn

Note: you may need to restart the kernel to use updated packages.


In [7]:
from imblearn.over_sampling import SMOTE

In [8]:
# Separate input features and target variable
x = df.drop('Class', axis=1)
y = df['Class']

In [9]:
# Split dataset into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [10]:
# Apply SMOTE to training set only
smote = SMOTE(random_state=42)
x_train_resampled, y_train_resampled = smote.fit_resample(x_train, y_train)

In [11]:
oversample = SMOTE()
x, y = oversample.fit_resample(x, y)

In [12]:
y.value_counts()

0    763
1    763
Name: Class, dtype: int64

In [13]:
x_dummy, x_test, y_dummy, y_test = train_test_split(x, y, test_size=0.33, random_state=42)

In [14]:
df2 = pd.concat([x, y], axis=1)
df2

Time        V1        V2        V3        V4        V5        V6  \
0        0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1        0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2        1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3        1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4        2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...    ...       ...       ...       ...       ...       ...       ...   
1521   508 -1.841084 -2.356072  2.361104  1.118158  2.323200  0.512718   
1522   532 -0.499531 -1.074169  0.626215  1.343492  0.343921 -1.066395   
1523   106  1.248683  0.341974  0.289050  0.668912 -0.328838 -0.972650   
1524   477 -2.149510 -1.654685  1.364293  1.398620  1.193539 -0.995590   
1525   532 -1.906887 -1.584297  2.009417  0.497592  1.910845  0.277431   

            V7        V8        V9  ...       V21       V22       V23  \
0     0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1    -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2     0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3     0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4     0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...        ...       ...       ...  ...       ...       ...       ...   
1521 -2.002555  0.739790  0.913560  ...  0.387667  1.096040  0.289520   
1522  0.188769 -0.152080 -0.102146  ...  0.100668 -0.307472  0.637649   
1523  0.070421 -0.174387  0.006444  ... -0.281485 -0.813513  0.125436   
1524  0.559469 -0.105321 -0.216505  ...  0.410093  0.370352  0.685435   
1525 -1.259411  0.556739  0.624351  ...  0.207191  0.699280  0.097695   

           V24       V25       V26       V27       V28      Amount  Class  
0     0.066928  0.128539 -0.189115  0.133558 -0.021053  149.620000      0  
1    -0.339846  0.167170  0.125895 -0.008983  0.014724    2.690000      1  
2    -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.660000      0  
3    -1.175575  0.647376 -0.221929  0.062723  0.061458  123.500000      0  
4     0.141267 -0.206010  0.502292  0.219422  0.215153   69.990000      0  
...        ...       ...       ...       ...       ...         ...    ...  
1521 -0.798487 -0.197999  0.823478 -0.119661 -0.161167    1.558297      1  
1522  0.082914  0.245093 -0.003558 -0.116393  0.032913  216.882305      1  
1523  0.272313  0.211125  0.097785 -0.021934  0.029294    2.690000      1  
1524 -0.006766  0.326844 -0.275736 -0.190088 -0.032501  305.857203      1  
1525 -0.635540 -0.260700  0.639072 -0.203470 -0.247046    1.321211      1  

[1526 rows x 31 columns]

## Generating samples

In [15]:
z = 1.96 
p = 0.5
e = 0.05 

In [16]:
from sklearn.preprocessing import normalize
from imblearn.over_sampling import RandomOverSampler
import math
import random
from sklearn.cluster import KMeans

#### 1. Simple  Random Sampling

In [17]:
ssize_srs = (z*z*p*(1-p))/(e*e)
ssize_srs = math.floor(ssize_srs)
print(ssize_srs)
def random(df, ssize_srs):
    random_sample = df.sample(n = ssize_srs)
    return(random_sample)
Sample = random(df2, ssize_srs)

x_ssize_srs = Sample.drop('Class', axis=1) #X_train
y_ssize_srs = Sample['Class'] #y_train
     

384


#### 2. Systematic Sampling

In [19]:
t = 3 #interval
ssize_systematic = df2.iloc[::t]
ssize_systematic

x_ssize_systematic = ssize_systematic.drop('Class', axis=1) #X_train
y_ssize_systematic = ssize_systematic['Class'] #y_train

#### 3. Stratified Sampling

In [20]:
s = 2 
ssize_stratified = (z*z*p*(1-p))/(e*e/s*s)
ssize_stratified = math.floor(ssize_srs)
print(ssize_stratified)
stratified = df2.groupby('Class')
stratified_df3 = stratified.apply(lambda x: x.sample(ssize_stratified))
stratified_df3
x_ssize_stratified = stratified_df3.drop('Class', axis=1) #X_train
y_ssize_stratified = stratified_df3['Class'] #y_train

384


#### 4. Cluster Sampling

In [21]:
kmeans = KMeans(n_clusters=10, random_state=42).fit(df2)
ssize_clustered = kmeans.labels_
cluster = [1,3,5,6,8]
cluster_series = pd.Series(ssize_clustered)
cluster_df = df2[cluster_series.isin(cluster)]
x_ssize_clustered = cluster_df.drop('Class', axis=1) #X_train
y_ssize_clustered = cluster_df['Class'] #y_train

#### 5. Snowball Sampling

In [22]:
def snowball(df, seed, n):
    snowball_sample = pd.DataFrame(columns=df.columns)
    snowball_sample = snowball_sample.append(df.loc[seed])
    current_sample = [seed]
    for i in range(n):
        rand_index = np.random.choice(len(current_sample))
        rand_observation = current_sample[rand_index]
        neighbors = df.loc[df['neighbor'] == rand_observation]
        new_neighbors = neighbors[~neighbors.index.isin(snowball_sample.index)]
        snowball_sample = snowball_sample.append(new_neighbors)
        current_sample.extend(new_neighbors.index.tolist())
    return snowball_sample
x_ssize_snowball = Sample.drop('Class', axis=1) #X_train
y_ssize_snowball = Sample['Class'] #y_train

## Training the models

In [23]:
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import BaggingClassifier

#### 1. Random Forest Classifier

In [24]:
def Random_forest_classifier(x_train, y_train, x_test, y_test):
  clf = RandomForestClassifier(n_estimators = 10)
  clf.fit(x_train, y_train)
  y_pred = clf.predict(x_test)
  return accuracy_score(y_test,y_pred)

#### 2. Bagging Classifier

In [25]:
def Bagging_classifier(x_train, y_train, x_test, y_test):
    base_classifier = DecisionTreeClassifier()
    bagging_classifier = BaggingClassifier(base_estimator=base_classifier, n_estimators=10)
    bagging_classifier.fit(x_train, y_train)
    y_pred = bagging_classifier.predict(x_test)
    return accuracy_score(y_test, y_pred)

#### 3. Decision Tree Classifier

In [26]:
def Decision_tree_classifier(x_train, y_train, x_test, y_test):
  tree = DecisionTreeClassifier(max_depth = 2).fit(x_train, y_train)
  tree_pred = tree.predict(x_test)
  return accuracy_score(y_test,tree_pred)

#### 4. Logistic Regression

In [27]:
def Logistic_regression(x_train, y_train, x_test, y_test):
  clf = LogisticRegression()
  clf.fit(x_train, y_train)
  y_pred = clf.predict(x_test)
  return accuracy_score(y_test,y_pred)

#### 5. K Neighbors Classifier

In [28]:
def KNN(x_train, y_train, x_test, y_test):
  knn = KNeighborsClassifier(n_neighbors = 10).fit(x_train, y_train)
  knn_pred = knn.predict(x_test)
  return accuracy_score(y_test,knn_pred)

## Calculating the accuracy for each sample type using each model

In [30]:
simple_random_Random_forest = Random_forest_classifier(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_bagging_classifier = Bagging_classifier(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_Decision_tree_classifier = Decision_tree_classifier(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_logistic = Logistic_regression(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_KNN = KNN(x_ssize_srs,y_ssize_srs,x_test,y_test)
simple_random_accuracy = [simple_random_Random_forest*100,simple_random_bagging_classifier*100,simple_random_Decision_tree_classifier*100,simple_random_logistic*100,simple_random_KNN*100]
simple_random_accuracy

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[98.4126984126984,
 96.62698412698413,
 90.87301587301587,
 91.07142857142857,
 73.80952380952381]

In [31]:
systematic_Random_forest = Random_forest_classifier(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_bagging_classifier = Bagging_classifier(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_Decision_tree_classifier = Decision_tree_classifier(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_logistic = Logistic_regression(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_KNN = KNN(x_ssize_systematic,y_ssize_systematic,x_test,y_test)
systematic_accuracy = [systematic_Random_forest*100,systematic_bagging_classifier*100,systematic_Decision_tree_classifier*100,systematic_logistic*100,systematic_KNN*100]
systematic_accuracy


/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[98.80952380952381,
 96.82539682539682,
 89.88095238095238,
 91.07142857142857,
 77.38095238095238]

In [32]:
stratified_Random_forest = Random_forest_classifier(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_bagging_classifier = Bagging_classifier(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_Decision_tree_classifier = Decision_tree_classifier(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_logistic = Logistic_regression(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_KNN = KNN(x_ssize_stratified,y_ssize_stratified,x_test,y_test)
stratified_accuracy = [stratified_Random_forest*100,stratified_bagging_classifier*100,stratified_Decision_tree_classifier*100,stratified_logistic*100,stratified_KNN*100]
stratified_accuracy

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[99.40476190476191,
 98.21428571428571,
 90.47619047619048,
 93.25396825396825,
 82.34126984126983]

In [36]:
clustered_Random_forest = Random_forest_classifier(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_bagging_classifier = Bagging_classifier(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_Decision_tree_classifier = Decision_tree_classifier(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_logistic = Logistic_regression(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_KNN = KNN(x_ssize_clustered,y_ssize_clustered,x_test,y_test)
clustered_accuracy = [clustered_Random_forest*100,clustered_bagging_classifier*100,clustered_Decision_tree_classifier*100,clustered_logistic*100,clustered_KNN*100]
clustered_accuracy
     

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[99.8015873015873,
 98.4126984126984,
 90.07936507936508,
 93.05555555555556,
 79.96031746031747]

In [37]:
snowball_Random_forest = Random_forest_classifier(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_bagging_classifier = Bagging_classifier(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_Decision_tree_classifier = Decision_tree_classifier(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_logistic = Logistic_regression(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_KNN = KNN(x_ssize_snowball,y_ssize_snowball,x_test,y_test)
snowball_accuracy = [snowball_Random_forest*100,snowball_bagging_classifier*100,snowball_Decision_tree_classifier*100,snowball_logistic*100,snowball_KNN*100]
snowball_accuracy

/Users/harshitapandey/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[97.22222222222221,
 96.42857142857143,
 90.87301587301587,
 91.07142857142857,
 73.80952380952381]

## Combining the results

In [38]:
final_accuracy = {
    'ML Model' : ['Random Forest','Bagging','Decision Tree','Logistic Regression','KNN'],
    'Simple Random Sampling(%age)' : simple_random_accuracy,
    'Stratified Sampling(%age)' : stratified_accuracy,
    'Cluster Sampling(%age)' : clustered_accuracy,
    'Systematic Sampling(%age)' : systematic_accuracy,
    'Snowball Sampling(%age)' : snowball_accuracy
}

final_accuracy_df = pd.DataFrame(final_accuracy)
final_accuracy_df

ML Model  Simple Random Sampling(%age)  \
0        Random Forest                     98.412698   
1              Bagging                     96.626984   
2        Decision Tree                     90.873016   
3  Logistic Regression                     91.071429   
4                  KNN                     73.809524   

   Stratified Sampling(%age)  Cluster Sampling(%age)  \
0                  99.404762               99.801587   
1                  98.214286               98.412698   
2                  90.476190               90.079365   
3                  93.253968               93.055556   
4                  82.341270               79.960317   

   Systematic Sampling(%age)  Snowball Sampling(%age)  
0                  98.809524                97.222222  
1                  96.825397                96.428571  
2                  89.880952                90.873016  
3                  91.071429                91.071429  
4                  77.380952                73.809524